File: 256/211-Lab05/radix_sort_html.py
from linked_queue import *

class RadixSortHTML(object):
    def __init__(self, filename: str) -> None:
        self.filename = filename
        self.buckets: list[LinkedQueue] = []
        self.tags: list[str] = []
        self.passes: int = 0

        self.initialize_buckets()

    def initialize_buckets(self) -> None:
        for i in range(0 , 27):
            self.buckets.append(LinkedQueue())

    def get_passes(self) -> int:
        return self.passes

    def read_tags(self) -> None:
        with open(self.filename , "r") as file:
            self.tags = file.read().splitlines()

    def process_tags(self) -> None:
        max: int = 0
        for string in self.tags:
            if len(string) > max:
                max = len(string)

        for i in range(len(self.tags)):
            while len(self.tags[i]) != max:
                self.tags[i] = self.tags[i] + " "

    def sort(self) -> None:
        max: int = len(self.tags[0]) - 1

        for pos in range(max , -1 , -1):
            self.count_sort(pos)

    def count_sort(self, position: int) -> None:
        for string in self.tags:
            letter: int = ord(string[position]) - ord('a')

            if letter == -65:
                self.buckets[26].enqueue(string)
            elif 0 <= letter <= 25:
                self.buckets[letter].enqueue(string)
            else:
                raise InvalidRadixSort("Invalid index")

        self.passes = 0
        for i in range(0 , 27):
            while not self.buckets[i].is_empty():
                self.tags[self.passes] = self.buckets[i].dequeue()
                self.passes = self.passes + 1

    def __str__(self) -> str:
        output: str = ""
        for tag in self.tags:
            output += tag + "\n"

        return output

File: 211/02-2025/Lab05/RadixSortHTML.java
// 02-26-2025

import java.io.File;
import java.io.FileNotFoundException;
import java.util.ArrayList;
import java.util.Scanner;

public class RadixSortHTML {
    private ArrayList<LinkedQueue<String>> buckets;
    ArrayList<String> tags;
    private String filename;
    private int passes;
    public RadixSortHTML(String filename) {
        this.filename = filename;
        this.buckets = new ArrayList<LinkedQueue<String>>();
        this.tags = new ArrayList<>();
        this.passes = 0;

        initializeBuckets();
    }
    public void initializeBuckets() {
        // 26 Letters AND a space
        for(int i = 0; i < 27; i++) {
            buckets.add(new LinkedQueue<>());
        }
    }

    public int getPasses() {
        return passes;
    }

    public void readTags() throws FileNotFoundException {
        Scanner scan = new Scanner(new File(filename)) ;

        while(scan.hasNextLine()) {
            tags.add(scan.nextLine()) ;
        }
    }

    public void processTags() throws FileNotFoundException {
        int max = 0 ;
        for(String str : tags) {
            if(str.length() > max) {
                max = str.length() ;
            }
        }

        for(int i = 0 ; i < tags.size() ; i++) {
            while(tags.get(i).length() != max) {
                tags.set(i , tags.get(i) + " ") ;
            }
        }
    }

    public void sort() throws FileNotFoundException {
        int max = tags.get(0).length() ;

        for(int pos = max - 1 ; pos >= 0 ; pos--) {
            countSort(pos) ;
        }
    }

    public void countSort(int position) {
        int n = tags.size() ;

        for(String str : tags) {
            int letter = str.charAt(position) - 'a' ;

            if(letter == -65) {
                buckets.get(26).enqueue(str) ;
            }
            else if(letter >= 0 && letter <= 25) {
                buckets.get(letter).enqueue(str) ;
            }
            else {
                throw new InvalidRadixSortException("invalid index") ;
            }
        }

        passes = 0 ;
        for(int i = 0 ; i <= 26 ; i++) {
            while(!buckets.get(i).isEmpty()) {
                tags.set(passes , buckets.get(i).dequeue()) ;
                passes++ ;
            }
        }
    }

    @Override  
    public String toString() {
        String output = "";
        for(String oneTag: tags)
            output += oneTag + "\n";
        return output;
    }
}
